staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go

```diff
@@ -62,10 +66,20 @@ func (r *result) Return() (runtime.Object, error) {
 	}
 }
 
+// PostTimeoutLoggerFunc is a function that can be used to log the result returned
+// by a ResultFunc after the request had timed out.
+// timedOutAt is the time the request had been timed out.
+// r is the result returned by the child goroutine.
+type PostTimeoutLoggerFunc func(timedOutAt time.Time, r *result)
+
 // FinishRequest makes a given ResultFunc asynchronous and handles errors returned by the response.
 func FinishRequest(ctx context.Context, fn ResultFunc) (runtime.Object, error) {
-	// the channel needs to be buffered to prevent the goroutine below from hanging indefinitely
-	// when the select statement reads something other than the one the goroutine sends on.
+	return finishRequest(ctx, fn, 30*time.Minute, logPostTimeoutResult)
```

Comment:
`30 minutes` - wow that's a long timeout, how did we end up with that number?

Reply:
instead of waiting infinitely, we are waiting for certain threshold. This is so that if there is such a goroutine that never returns we would know about it.  I am adding this to clarify:
```
const (
	// how much time the post-time logger goroutine will wait for the sender
	// (child goroutine executing ResultFunc) to send a result after the request.
	// had timed out.
	postTimeoutLoggerWait = 30*time.Minute
)
```